### ISSUE TITLE
Proposal: allow open types in nameof

### ISSUE THREAD

**Issue by jskeet** [dotnet/roslyn#15983]
(Created from #4827)

Currently to use `nameof` with a generic type, a type argument needs to be used. Problems:

- It's very odd to require something to be specified within an operand when it has *no* impact on the result
- It's inconsistent with `typeof`
- It means that changing a constraint on a type parameter can break uses of `nameof` entirely unnecessarily

I propose that the exact same syntax as `typeof` is used. So for example:

```csharp
class GenericClass<T>
{
    public string X { get; }
}

class GenericClass<T1, T2>
{
    public string Y { get; }
}
```

These expressions are currently valid, and would stay valid:

```csharp
nameof(GenericClass<int>)
nameof(GenericClass<int>.X)
nameof(GenericClass<int, byte>)
nameof(GenericClass<int, byte>.Y)
```

These expressions would *become* valid, and would have the same respective results:

```csharp
nameof(GenericClass<>)
nameof(GenericClass<>.X)
nameof(GenericClass<,>)
nameof(GenericClass<,>.Y)
```

**Comment by alrz** [comment]
Should `nameof(GenericClass<>)` return `GenericClass`1"?

**Comment by jnm2** [comment]
> Should `nameof(GenericClass<>)` return "`GenericClass`1`"?

Absolutely not. The main use case for `nameof(GenericClass<>)` is to follow the name `GenericClass` around even if you rename-refactor `GenericClass<>` without renaming `GenericClass` or `GenericClass<,>`.

**Comment by alrz** [comment]
A `GenericClass<,>` is different from a `GenericClass<>`; if you change arity, the latter would not be the same type. In fact, `System.Type.Name` for generic types follows the above pattern as they are not the same.

**Comment by eyalsk** [comment]
@alrz It should return a _similar_ thing to `typeof(...)` so `typeof(IList<>).ToString()` would return ``System.Collections.Generic.IList`1[T]`` so I guess that `nameof((IList<>)` should return ``IList`1[T]``.

**Comment by jnm2** [comment]
@alrz I don't disagree, but the important part, what `nameof` what created for, is not the CLR mangled type name but rather the C# language name.

**Comment by eyalsk** [comment]
@jnm2 Yeah you got a point. :D

**Comment by alrz** [comment]
>not the CLR mangled type name but rather the C# language name

That could be the case. Not sure about nameof, but https://github.com/dotnet/roslyn/issues/10972 certainly works for me.

**Comment by jnm2** [comment]
@alrz If you wanted the CLR type name, `typeof(GenericClass<>).Name` has always been a handy tool for that.

**Comment by eyalsk** [comment]
@jnm2 Yes, only it's not a constant.

**Comment by jnm2** [comment]
That's right, but it could be if Roslyn starts recognizing `typeof(x).Name`. Even so it would not have the same uses cases as `nameof`.

**Comment by jskeet** [comment]
As per the original proposal, I'm suggesting that using the open type should produce the same results as a constructed type. While I can see there being uses for the generic form, we can't change the results of `nameof` with the constructed type now, and it would be highly inconsistent for the open type to give a different result.

**Comment by paulomorgado** [comment]
However, I'm pretty comfortable with using constructed types, as I'm already used to it.

**Comment by alrz** [comment]
I think using discards would make the implementation easier as there would be no syntax changes.
```cs
nameof(GenericClass<_>)
nameof(GenericClass<_>.X)
nameof(GenericClass<_,_>)
nameof(GenericClass<_,_>.Y)
```

**Comment by eyalsk** [comment]
@alrz Shouldn't this work too then? `typeof(IDictionary<_>)`?

**Comment by alrz** [comment]
@eyalsk It could, but it does not resolve any ambiguity because there is none. `typeof` is a keyword so the compiler wouldn't have any problem to permit a special "open generic" syntax in that context, but `nameof` on the other hand, is a contextual keyword as it parsed as an invocation expression. Not that it wouldn't be possible but I think that it requires syntax changes. It could take advantage of `_` being an identifier, just like it has been done for type patterns.

**Comment by eyalsk** [comment]
@alrz I understand but it still feels a bit odd that we need to use commas (_or underscores_) at one place and (_only_) underscores at another.

**Comment by jnm2** [comment]
I really dislike the underscores. I see no reason to make the syntax any different than it is for `typeof`.

**Comment by gafter** [comment]
@jnm2 
>  I see no reason to make the syntax any different than it is for `typeof`.

Can you please outline the language grammar you have in mind?

**Comment by jnm2** [comment]
@gafter Precisely what @jskeet listed above. It retains perfect consistency with the rest of C#, which is the main motivating factor here in addition to ease of use:

> These expressions are currently valid, and would stay valid:
> 
> ```c#
> nameof(GenericClass<int>)
> nameof(GenericClass<int>.X)
> nameof(GenericClass<int, byte>)
> nameof(GenericClass<int, byte>.Y)
> ```
> 
> These expressions would become valid, and would have the same respective results:
> 
> ```c#
> nameof(GenericClass<>)
> nameof(GenericClass<>.X)
> nameof(GenericClass<,>)
> nameof(GenericClass<,>.Y)
> ```

No underscores.

**Comment by gafter** [comment]
@jnm2 Sorry, @jskeet didn't outline a language grammar. If you don't have any idea either, I'll let the issue sit.

**Comment by gafter** [comment]
To be clear, the problem is that `nameof` is treated as an invocation expression for purposes of syntax analysis. And an invocation expression simply cannot have an open type in it.

**Comment by jnm2** [comment]
@gafter I am not familiar with language grammar, but it seems obvious to me that `nameof` will have to be treated as an entirely new hybrid class of expression.

**Comment by gafter** [comment]
@jnm2 I agree, that would work. But we treat it as an invocation expression so we do not break backward compatibility.

**Comment by jnm2** [comment]
I want this badly enough that I'll get up to speed on language grammar to help get this through, but I doubt I'm the most valuable person in this thread for that purpose. :')

What if the hybrid expression, the argument to `nameof`, could be an InfoOf expression in reference to the infamous `infoof` operator we've talked about for years. For now it could only be used as an argument to `nameof` but this opens doors for the future.

`List<>.Count` would be the spiritual equivalent of `typeof(List<>).GetProperty("Count")`, and `List<int>.Count` would be the spiritual equivalent of `typeof(List<int>).GetProperty("Count")`. It wouldn't make a difference to `nameof` which one you write, because it's getting the C# language name (not the CLR name) of `Count` which does depending on the type parameter.

ReSharper would suggest that you remove the closed type parameter if it's used for `nameof`. However, if in the future we add `infoof`, there would be a real difference between the two members. Similar to the difference `MakeGenericType` and `MakeGenericMethod` make.

> But we treat it as an invocation expression so we do not break backward compatibility.

At a source level, any currently-valid argument to `nameof` would remain valid. Can you help be understand the compatibility issue in more detail? Is it just a Roslyn API back-compat issue?

**Comment by jnm2** [comment]
https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#nameof-expressions

Okay, so the operand to `nameof`, what is currently called the **named_entity** expression, would be what I was calling the InfoOf expression.
And if you took to the `info_of` expression idea, this sentence would be removed:

> It is a compile-time error for a named_entity designating a method group to have a type_argument_list.

**Comment by jnm2** [comment]
To get back to basics here, this is what we currently have:

```antlr
named_entity
    : simple_name
    | named_entity_target '.' identifier type_argument_list?
    ;
```

What if `type_argument_list` was simply changed to allow both open and closed types? (Looking for it's name...)

**Comment by jnm2** [comment]
So there would be a version of `named_entity` but with [`generic_dimension_specifier`](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#the-typeof-operator) instead of [`type_arguments`](https://github.com/dotnet/csharplang/blob/master/spec/types.md#type-arguments).

**Comment by CyrusNajmabadi** [comment]
A simple way to handle this from the grammar would be to just change:

```
simple-name:
    identifier   type-argument-list
    identifier   generic-dimension-specifier
```

With normative text to specify that the latter are only legal in the respective locations in typeof/nameof.  

--

This grammar would be essentially mimic what we are already doing in the C# parser today.  i.e we support "generic-dimension-specifiers" everywhere, we just report errors if you use them in inappropriate locations.

**Comment by gafter** [comment]
The spec grammar is ambiguous, because it requires semantic context to distinguish a `typeof` operator from a method invocation of the same name. The compiler simply parses an invocation expression. So it is the method invocation grammar that needs to be adjusted.

**Comment by jnm2** [comment]
@CyrusNajmabadi @gafter Do you need any more guesswork from me or do you have a handle on it? :)

**Comment by craigajohnson** [comment]
Absolutely yes please to this feature and please include VB.

**Comment by jnm2** [comment]
@jskeet Would you please port this to csharplang?

**Comment by jskeet** [comment]
@jnm2: Done - follow the link above :)
